Adds Murex shell support

JeffMelton opened the pull request:
Murex is niche, but I use it, and I miss zoxide. 😄

<!-- gk-ai-analysis-start:57b10feec66d969ca72a3f2179a7f2b9f7bac941 -->
<!-- gk-ai-analysis-data:eyJzdW1tYXJ5IjoiQWRkcyBpbml0aWFsaXphdGlvbiBhbmQgY29tbWFuZCBzdXBwb3J0IGZvciB0aGUgTXVyZXggc2hlbGwuIiwia2V5SW5zaWdodHMiOlt7InRpdGxlIjoiTXVyZXggdGVtcGxhdGUgYWRkZWQiLCJkZXNjcmlwdGlvbiI6IkludHJvZHVjZXMgdGhlIE11cmV4IHNoZWxsIGluaXRpYWxpemF0aW9uIHNjcmlwdCB0ZW1wbGF0ZSwgZGVmaW5pbmcgZnVuY3Rpb25zIGxpa2UgYF9fem94aWRlX3pgLCBgX196b3hpZGVfemlgLCBhbmQgYF9fem94aWRlX2NkYC4iLCJzZXZlcml0eSI6Im1lZGl1bSIsImZpbGVQYXRoIjoidGVtcGxhdGVzL211cmV4LnR4dCIsImxpbmVTdGFydCI6MX0seyJ0aXRsZSI6Ik11cmV4IGludGVncmF0ZWQgYXMgYSBzdXBwb3J0ZWQgc2hlbGwiLCJkZXNjcmlwdGlvbiI6IkFkZHMgTXVyZXggdG8gdGhlIGBJbml0U2hlbGxgIGVudW1lcmF0aW9uIGFuZCB3aXJlcyBpdCBpbnRvIHRoZSBzaGVsbCBpbml0aWFsaXphdGlvbiBhbmQgdGVtcGxhdGUgcmVuZGVyaW5nIHBpcGVsaW5lLiIsInNldmVyaXR5IjoibG93IiwiZmlsZVBhdGgiOiJzcmMvY21kL2luaXQucnMiLCJsaW5lU3RhcnQiOjZ9LHsidGl0bGUiOiJNdXJleCB0ZXN0IGNvdmVyYWdlIGFkZGVkIiwiZGVzY3JpcHRpb24iOiJBZGRzIHNuYXBzaG90LXN0eWxlIHRlbXBsYXRlIHJlbmRlcmluZyB0ZXN0cyBhbmQgcGFyYW1ldGVyIGNoZWNrcyBmb3IgTXVyZXggdG8gZW5zdXJlIGhvb2tzLCBhbGlhc2VzLCBhbmQgZGlyZWN0IGNkIGJlaGF2aW9ycyBhcmUgY29ycmVjdGx5IGdlbmVyYXRlZC4iLCJzZXZlcml0eSI6ImxvdyIsImZpbGVQYXRoIjoic3JjL3NoZWxsLnJzIiwibGluZVN0YXJ0IjoyNTB9XSwiaXNzdWVzIjpbXSwic3VnZ2VzdGlvbnMiOltdLCJzZWN1cml0eSI6W119 -->
<!-- gk-ai-analysis-end:57b10feec66d969ca72a3f2179a7f2b9f7bac941 -->

<!-- gitkraken-review-badge-begin -->
---
<a href="https://gitkraken.dev/review/github/ajeetdsouza/zoxide/pull/1111?source=pr_review_chip">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://gitkraken.dev/images/figures/gitkraken-review-badge-dark.svg">
    <img src="https://gitkraken.dev/images/figures/gitkraken-review-badge-light.svg" alt="Open with GitKraken">
  </picture>
</a>
<!-- gitkraken-review-badge-end -->

JeffMelton commented:
Sure can! Thanks for reviewing. 😁

On Sep 7, 2025, at 12:55 AM, ajeetdsouza ***@***.***> wrote:

﻿
[https://avatars.githubusercontent.com/u/1777663?s=20&v=4]ajeetdsouza left a comment (ajeetdsouza/zoxide#1111)<https://github.com/ajeetdsouza/zoxide/pull/1111#issuecomment-3263510932>

Hi, I've added some minor bugfixes and made it more consistent with the other shells. There's a few more things I'd like to fix before merging:

  *   PWD hooks are possible to implement here. What you need to do is use a global variable called __zoxide_oldpwd, and set up a prompt hook that compares __zoxide_oldpwd with $PWD. If it has changed, set __zoxide_oldpwd=$PWD and call zoxide add $PWD (since the directory has changed).
  *   z (with no arguments) does not work.
  *   If you zoxide has no result for your query (e.g. z aoiejfaoiejf), you get a full error trace with a line number. Ideally, we should just return 1 in such a situation.

Do you think you could work on these?

—
Reply to this email directly, view it on GitHub<https://github.com/ajeetdsouza/zoxide/pull/1111#issuecomment-3263510932>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AABIO2C63LIADNU7XDC2CXD3RPCFPAVCNFSM6AAAAACF2LCEW2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTENRTGUYTAOJTGI>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

JeffMelton commented:
All right, @ajeetdsouza, thanks for the feedback. I think I've covered your requests. Tests pass here, but I see the ubuntu CI test is still failing. I don't have `nix-shell` set up, but I'm trying to replicate as much as I can in a container environment. Do please let me know if you'd like to see anything else.

JeffMelton commented:
Ah, I've found a flaw in my implementation. I'll update again here once I think I've sorted it.

ajeetdsouza commented:
Tried `zoxide init murex -> source` with murex v7.0.2107, but the hook does not seem to be working. A way to see if the hook is working is to do something like this:

```shell
zoxide query --score /tmp
z /tmp
# the score should have increased
zoxide query --score /tmp
# when in the same directory, the score should stay the same
zoxide query --score /tmp
```

Also, `z` with no arguments still doesn't seem to work for me:

```
Error in `__zoxide_z` (<stdin> 1,2637):
      Command: __zoxide_z  @PARAMS
      Error: array '@PARAMS' is empty
```

JeffMelton commented:
All right, thanks. I'll keep poking at it.

JeffMelton commented:
Okay, thanks, that does clarify. Mine is "working"/updating the score because it's overriding default, which means the PWD hook still isn't working right.

JeffMelton committed:
Add Murex shell support

Adds the Murex shell (https://murex.rocks) to the set of shells
supported by 'zoxide init'.

Changes:
- templates/murex.txt: new init script template implementing
  PWD/Prompt hooks via 'event onPrompt name=command-completion',
  __zoxide_z/zi helpers with empty-PARAMS handling, and clean
  return-1 on no-match queries.
- src/shell.rs: wires the template + two tests:
  * murex_murex (functional, runs the rendered init under murex -c),
    parameterized across all opts combinations and gated by the
    nix-dev feature so it only runs in the project's nix-shell
  * murex_template_has_direct_path_handling (string assertions on
    the rendered template, runs always)
- src/cmd/cmd.rs: adds Murex to the InitShell enum
- src/cmd/init.rs: imports Murex and adds the match arm
- shell.nix: pins Murex to v7.2.1001 via a surgical overrideAttrs
  on the existing pkgs.murex (only this single package changes; the
  nixpkgs pin itself stays put). v7.0 introduced the
  'command-completion' onPrompt interrupt that the prompt/PWD hooks
  depend on, so the version bundled with the current nixpkgs pin
  (v6.4.2063) is too old to source the rendered init.

Completion stubs in contrib/completions/ are not touched directly;
they regenerate via build.rs when the InitShell enum changes.

Closes #1085

JeffMelton commented:
Hey @ajeetdsouza — circling back on this with what I hope is the last meaningful update before merge review. No rush on your end; just consolidating everything in one place so when you do have time it's a short read.

## TL;DR

- `murex -c <source>` exits non-zero on broken init. Empirically proven (transcript below).
- All four behaviors you flagged previously now work cleanly: PWD hook, `z` (no args), `z <no-match>`, `zi` (no args).
- One small Nix-side change: pinned Murex to `v7.2.1001` via a surgical `overrideAttrs` (only this package, nixpkgs pin untouched). Background and rationale below.
- 525/525 tests + lint green.

## Behavioral demos

All transcripts captured against `murex v7.2.1001`, rendered init sourced cleanly, no warnings.

### PWD hook (your original repro)

```
tmp » zoxide query --score /tmp
zoxide: no match found
tmp » cd /tmp
tmp » zoxide query --score /tmp
   4.0 /tmp
```

Score increments correctly on directory change via the default `--hook=pwd` path. The hook fires from a `event onPrompt name=command-completion` handler (v7.0+ feature — more on that below).

### `z` with no arguments

```
tmp » z
~ » pwd
/root
```

The wrapper now checks `$PARAMS` arity (as JSON) before expanding `@PARAMS`, since expanding an empty `@PARAMS` errors before the function body runs. This was the bug behind your `array '@PARAMS' is empty` trace.

### `z <no-match>`

```
tmp » z aoiejfaoiejf
tmp »
```

Silent return 1, no stack trace.

### `zi` with no arguments

Fzf launches against the current DB; selection cd's. Same arity-handling fix had to be applied inside `__zoxide_zi` itself — the existing `trypipe <!null>` was swallowing the `@PARAMS` error and quietly suppressing fzf launch.

## Merge gate: `murex -c` exits non-zero on broken init

This was your stated bar for merging. To prove it directly, I broke the template in two different ways and ran the existing `murex_murex` test suite:

**Test 1 — invalid event interrupt name:**
```
$ sed -i 's/command-completion/not-a-real-interrupt/' templates/murex.txt
$ cargo nextest run --features nix-dev shell::tests::murex_murex
...
Summary: 24 tests failed, all murex_murex_* variants
```

**Test 2 — outright syntax error:**
```
$ echo 'function {{{ }}}' >> templates/murex.txt
$ cargo nextest run --features nix-dev shell::tests::murex_murex
...
Summary: 24 tests failed
```

So the existing `murex -c <source>` exit-code check catches both syntax errors and runtime registration errors during sourcing. That's broader than the strict "syntax error" framing in your earlier comment — anything that fails at source-time trips it.

(Bug-honesty caveat below.)

## The Nix pin bump (and why it's safe)

Working through this, I hit a snag: the rendered init uses `event onPrompt name=command-completion`, which is a Murex **v7.0+** feature. The `nixpkgs` revision pinned by `shell.nix` ships `murex v6.4.2063`, where `command-completion` isn't a registered interrupt, so the init errored during source under CI.

You're on v7.0.2107 per your September comment, so this isn't a new user-facing constraint — it just brings the CI environment in line with what real users have.

I tried two approaches:

1. **Bump the whole nixpkgs pin** — too invasive, ripples through every other shell.
2. **Surgical override on just the `murex` derivation** — preferred. The file already uses the `pkgs.murex.overrideAttrs (old: { doCheck = false; })` idiom, so I just extended it:

```nix
(pkgs.murex.overrideAttrs (old: rec {
  version = "7.2.1001";
  src = pkgs.fetchFromGitHub {
    owner = "lmorg";
    repo = "murex";
    rev = "v${version}";
    hash = "sha256-Ua5KEtT1HXRCqW4MwB0dYCd03DBrliEfgiSmcp+vZS8=";
  };
  vendorHash = "sha256-MaBBi2Qi7s9lfRWmnYkyr7PtwzC7ZL0jmyUXzISOXVg=";
  doCheck = false;
}))
```

One package, one override, all other shells unaffected. v7.2.1001 is the current stable release.

If you'd rather not bump Murex and prefer a v6-compatible template, the alternative is to drop `event onPrompt name=command-completion` and use an older hook mechanism (likely `before`/`after` interrupts), at the cost of more complex hook logic. Happy to do that instead if you'd prefer — just let me know.

## Honest caveats

A few things worth surfacing before you spend time on this:

1. **The test's coverage is shallower than ideal.** It catches errors that fire *during source*. It does **not** catch runtime issues at the call site — for example, the empty-`@PARAMS` bug you saw originally would still slip past the test, because `__zoxide_z` doesn't get invoked just by sourcing. I caught the remaining instances of that pattern interactively. If you want stronger coverage, we could extend the test to actually invoke `z` and `zi` after sourcing, but that's incremental work and I didn't want to scope-creep this PR.

2. **The Nix startup warning you mentioned** — I poked at this. Inside `nix-shell --pure`, fzf's preview command (`command -p ls ...`, set by `src/util.rs`) fails because there's no `/bin/ls` in pure Nix. This affects **every** shell init under nix-shell (bash, zsh, fish, all of them), not just Murex. It's an artifact of the CI environment, not a Murex-specific issue. In a normal Linux/macOS environment it works fine.

3. **The Murex startup warning under Nix** that you specifically noted in your Nov comment — I wasn't able to reproduce a warning at startup under v7.2.1001 in the project's nix-shell. If you can share the exact warning text you saw, I'm happy to dig further.

## What changed since the last review

- `templates/murex.txt` — empty-`@PARAMS` handling in both the `z`/`zi` wrappers and inside `__zoxide_zi`; comments explaining the arity-branch idiom.
- `shell.nix` — surgical Murex pin to v7.2.1001 as described above.
- Commit signed under my personal identity (jeff@melton.cloud).

Single commit (`Add Murex shell support`), rebased clean on master.

Ready for re-review whenever you have bandwidth — no rush. Thanks for the patience on this one.
